sql: add parse_timestamp builtin

RaduBerinde opened the pull request:
Add a builtin that can be used to parse timestamp strings. This is
like a cast, but it does not accept relative timestamps so it can be
immutable.

Only immutable expressions are allowed in computed column expressions
or partial index predicates; unlike casts, the new function can be
used in such expressions.

Fixes #60578.

Release notes (sql change): A new parse_timestamp function can be used
to parse absolute timestamp strings in computed column expressions or
partial index predicates.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60772)
<!-- Reviewable:end -->

otan commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -2732,6 +2732,31 @@ may increase either contention or retry errors, or both.`,
+					return nil, err
+				}
+				if dependsOnContext {
+					return nil, errors.New("relative timestamps are not supported")
```
mind putting a `pgerrorNewf` & code on this?

otan commented on pkg/sql/logictest/testdata/logic_test/datetime:
```diff
@@ -1703,3 +1703,35 @@ query TT
+INSERT INTO timestamps VALUES ('2020-01-02 01:02:03'), ('2015-08-25 04:45:45.53453+01:00'), (NULL)
+
+query TT colnames
+SELECT * FROM timestamps ORDER BY s
```
mind setting local time zone to say NY and checking the time values are the same?

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-:-MTvs-fBEI_EM_M6Y8yX:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/sem/builtins/builtins.go, line 2735 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-MTvr6qBDmUUvbw2YUa5:-MTvr6qC8Vid2LvWw6rG:bgv0xlr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1c3532528964fc5fb203c6dc982fbffdd23b48f/pkg/sql/sem/builtins/builtins.go#L2735)):*
> ```Go
> 	),
> 
> 	// parse_timestamp is converts strings to timestamps. It is useful in
> ```

[nit] is converts -> converts


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: add parse_timestamp builtin

Add a new builtin function that can be used to parse timestamp
strings. This is like a cast, but it does not accept relative
timestamps so it can be immutable.

Only immutable expressions are allowed in computed column expressions
or partial index predicates; unlike casts, the new function can be
used in such expressions.

Fixes #60578.

Release notes (sql change): A new parse_timestamp function can be used
to parse absolute timestamp strings in computed column expressions or
partial index predicates.

RaduBerinde reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-:-MTvx9Na-zmLgSPRBZGI:bz5mair)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @otan, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/datetime, line 1729 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-MTsNlV3ssr0Vomb6QMv-r1-1729:-MTvx-9gCD6YuPAIAgy7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1c3532528964fc5fb203c6dc982fbffdd23b48f/pkg/sql/logictest/testdata/logic_test/datetime#L1729)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

mind setting local time zone to say NY and checking the time values are the same?
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/builtins.go, line 2735 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-MTvr6qBDmUUvbw2YUa5:-MTvwT8S7zLoyr9yaS9j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1c3532528964fc5fb203c6dc982fbffdd23b48f/pkg/sql/sem/builtins/builtins.go#L2735)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] is converts -> converts
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/builtins.go, line 2751 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-MTsNlV3ssr0Vomb6QMw-r1-2751:-MTvwLSA0XWJ5GMBWAgo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1c3532528964fc5fb203c6dc982fbffdd23b48f/pkg/sql/sem/builtins/builtins.go#L2751)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

mind putting a `pgerrorNewf` & code on this?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60772)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @otan)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-:-MTwuip3DaBoEtHN35KG:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @RaduBerinde)

___
*[pkg/sql/sem/builtins/builtins.go, line 2757 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-MTwuSlz-8OgsF6Tlk1S:-MTwuSlz-8OgsF6Tlk1T:b1zwxct) ([raw file](https://github.com/cockroachdb/cockroach/blob/e711b72ccdc0f3dfdd75ec3183cdd59eb319eadd/pkg/sql/sem/builtins/builtins.go#L2757)):*
> ```Go
> 				return ts, nil
> 			},
> 			Info:       "Convert a string containing an absolute timestamp to the corresponding timestamp.",
> ```

Is it worth mentioning that the default TZ is UTC if the string does not contain a TZ? If not here, maybe in the comment above?


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/sem/builtins/builtins.go, line 2757 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-MTwuSlz-8OgsF6Tlk1S:-MTzS_7R6SMdinrg3L4E:bvfiyos) ([raw file](https://github.com/cockroachdb/cockroach/blob/e711b72ccdc0f3dfdd75ec3183cdd59eb319eadd/pkg/sql/sem/builtins/builtins.go#L2757)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Is it worth mentioning that the default TZ is UTC if the string does not contain a TZ? If not here, maybe in the comment above?
</blockquote></details>

If should be clear given that it's a timestamp (and not timestamptz). I think even if it has a timezone, it is ignored. It's just like a cast (though I agree we should document those somewhere in more detail).
Maybe it should have been called parse_utc_timestamp..


<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/sem/builtins/builtins.go, line 2757 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60772#-MTwuSlz-8OgsF6Tlk1S:-MU-nvPi7jloDhzRkLMA:bsf2jkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e711b72ccdc0f3dfdd75ec3183cdd59eb319eadd/pkg/sql/sem/builtins/builtins.go#L2757)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

If should be clear given that it's a timestamp (and not timestamptz). I think even if it has a timezone, it is ignored. It's just like a cast (though I agree we should document those somewhere in more detail).
Maybe it should have been called parse_utc_timestamp..
</blockquote></details>

I think the name is fine - `parse_utc_timestamp` is misleading IMO because it implies there is such thing as a UTC-based timestamp, but in reality no timestamp has any timezone designation. It might make sense to document that any TZ information in the string is ignored, but I see how it should be clear from the name.

I think the term "absolute timestamp" confuses me because by definition no timestamp is absolute - it's always dependent on some reference timezone. A timestamp is never a single absolute point in time, while a timestamptz is.

As an aside, it's also confusing that we display timestamps with a time zone of +00. There is no timezone, so displaying +00 is incorrect. Even if we store the timestamp in the same way as a timestamptz with a UTC time zone, we shouldn't display any timezone to the user.

```
defaultdb> select parse_timestamp('2020-01-21T12:00:00');
        parse_timestamp
--------------------------------
  2020-01-21 12:00:00+00:00:00
```

But now I'm just spiraling...


<!-- Sent from Reviewable.io -->

otan commented:
> But now I'm just spiraling...

hate spiralling?
migrate us from lib/pq to pgx today (from the CLI) and this all goes away :D
